proof/refine/ARM/Finalise_R.thy

```diff
@@ -2326,6 +2326,327 @@ lemma replyClear_invs'[wp]:
+  unfolding replyPop_def
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)+
+  apply (rule hoare_seq_ext)
+   apply (wpsimp wp: cleanReply_list_refs_of_replies')
+  apply (rule hoare_when_cases)
+   apply (clarsimp simp: obj_at'_def)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp simp: comp_def\<close>, simp?)+
+  apply (rule hoare_seq_ext)
+   apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)
+   apply (wpsimp wp: hoare_when_cases)
+  apply (rule hoare_when_cases[rotated])
+   apply (wpsimp wp: updateReply_list_refs_of_replies')
+   apply (rename_tac prevReplyPtr prevReply)
+   apply (frule_tac reply=prevReply in ko_at'_replies_of')
+   apply (frule (3) sym_refs_replyNext)
+   apply (clarsimp,
+          erule delta_sym_refs
+          ; clarsimp simp: isHead_def obj_at'_def projectKOs list_refs_of_reply'_def
+                           list_refs_of_replies'_def opt_map_def get_refs_def
+                    split: if_splits option.splits)+
+  done
+
+lemma replyPop_iflive:
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> valid_objs' s \<and>
+        ex_nonz_cap_to' tcbPtr s\<rbrace>
+   replyPop replyPtr tcbPtr
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  unfolding replyPop_def
+  supply if_split[split del]
+  apply (wpsimp wp: schedContextDonate_if_live_then_nonz_cap' hoare_vcg_if_lift2 hoare_vcg_conj_lift
+                    hoare_vcg_imp_lift' threadGet_const)
+                  apply (wp updateReply_obj_at'_only_st_qd_ft)
+                 apply (wpsimp wp: updateReply_iflive'_strong replyNext_update_valid_objs')
+                apply (wpsimp wp: replyNext_update_valid_objs' hoare_vcg_imp_lift')+
+                 apply (wpsimp wp: updateReply_iflive'_strong replyNext_update_valid_objs')
+                apply (wpsimp wp: hoare_vcg_imp_lift')
+               apply (wpsimp wp: hoare_vcg_if_lift2 hoare_vcg_conj_lift hoare_vcg_imp_lift'
+                                 hoare_vcg_disj_lift hoare_vcg_all_lift)
+                 apply (wp updateReply_obj_at'_only_st_qd_ft)
+                apply (wpsimp wp: replyNext_update_valid_objs')
+               apply (wp updateReply_iflive'_strong)
+              apply (wpsimp wp: hoare_vcg_if_lift2 hoare_vcg_conj_lift hoare_vcg_imp_lift'
+                                hoare_vcg_disj_lift hoare_vcg_all_lift)
+             apply (wpsimp wp: gts_wp')+
+  apply (clarsimp simp: isHead_to_head split: if_splits)
+  apply (intro conjI | clarsimp)+
+          apply auto
+              prefer 7
+              apply (erule if_live_then_nonz_capE')
+              apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKOs live_sc'_def)
+             apply (erule if_live_then_nonz_capE')
+             apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKOs live_sc'_def)
+  sorry
```

Comment:
@MitchellBuckley this is the lemma we discussed you should take a look at

Reply:
I believe that Michael took a look at this. I suppose his work should be cherry-picked on top here?